processing/src/main/java/org/apache/druid/query/filter/TypedInFilter.java

```diff
@@ -0,0 +1,742 @@
+import org.apache.druid.common.config.NullHandling;
+import org.apache.druid.error.InvalidInput;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.math.expr.Evals;
+import org.apache.druid.query.cache.CacheKeyBuilder;
+import org.apache.druid.query.filter.vector.VectorValueMatcher;
+import org.apache.druid.query.filter.vector.VectorValueMatcherColumnProcessorFactory;
+import org.apache.druid.segment.ColumnInspector;
+import org.apache.druid.segment.ColumnProcessors;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.DimensionHandlerUtils;
+import org.apache.druid.segment.column.ColumnIndexSupplier;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.NullableTypeStrategy;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.segment.index.BitmapColumnIndex;
+import org.apache.druid.segment.index.semantic.Utf8ValueSetIndexes;
+import org.apache.druid.segment.index.semantic.ValueSetIndexes;
+import org.apache.druid.segment.vector.VectorColumnSelectorFactory;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Approximately like the SQL 'IN' filter, with the main difference being that this will match NULL values if contained
+ * in the values list instead of ignoring them.
+ * <p>
+ * This is a typed version of {@link InDimFilter}, which allows the match values to exist in their native type and
+ * sorted in their type native order for better performance matching against all column types.
+ */
+public class TypedInFilter extends AbstractOptimizableDimFilter implements Filter
+{
+  /**
+   * Column to match {@link #sortedMatchValues} or {@link #sortedUtf8MatchValueBytes} against.
+   */
+  private final String column;
+
+  /**
+   * Type of values contained in {@link #sortedMatchValues}. This might be the same or different than the
+   * {@link ColumnType} of {@link #column}, but is encouraged to be the same there are several optimizations available
+   * if they match.
+   */
+  private final ColumnType matchValueType;
+
+  /**
+   * Unsorted values. This will be null if the values are found to be sorted, or have been already sorted "upstream".
+   * Otherwise, this set of values will be lazily computed into {@link #sortedMatchValues} as needed, e.g. for
+   * JSON serialization, cache key building, building a hashcode, or checking equality.
+   */
+  @Nullable
+  private final List<?> unsortedValues;
+
+  /**
+   * Supplier for list of values sorted by {@link #matchValueType}. This is lazily computed if
+   * {@link #unsortedValues} is not null and previously sorted.
```

Comment:
would be useful to include a comment about whether this can contain duplicates or not, and if duplicates are not meant to be here, then also include a test that validates that deduplication happens on `values` (even if they are provided in sorted order).